docs/datatypes/tsemaphore.md

```diff
@@ -0,0 +1,99 @@
+---
+id: datatypes_tsemaphore
+title:  "TSemaphore"
+---
+`TSemaphore` is a semaphore with transactional semantics that can be used to control access to a common resource. It 
+holds a certain number of permits, and permits may be acquired or released.
+
+## Create a TSemaphore
+Creating a `TSemaphore` with 10 permits:
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tSemaphoreCreate: STM[Nothing, TSemaphore] = TSemaphore.make(10L)
+```
+
+## Acquire a permit
+Acquiring a permit reduces the number of remaining permits that the `TSemaphore` contains. Acquiring a permit is done 
+when a user wants to access a common shared resource:
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tSemaphoreAcq: UIO[TSemaphore] = (for {
```

Comment:
This might be a little more readable if you pulled the transaction and committing it into separate expressions so you didn't have the awkward `(for` syntax:

```scala
val transaction = ???
transaction.commit
```